Ibind hw

robertis93 opened the pull request:
bindservice

robertis93 committed:
with contactList

robertis93 committed:
Merge branch 'main' into contact-list

# Conflicts:
#	.idea/misc.xml
#	.idea/modules.xml

robertis93 committed:
without .idea

robertis93 committed:
without recyclerview and with viewbinding

robertis93 committed:
newinsyance

robertis93 committed:
bindservice

elestatte87 commented on app/build.gradle:
```diff
@@ -42,4 +49,8 @@ dependencies {
     testImplementation 'junit:junit:4.+'
     androidTestImplementation 'androidx.test.ext:junit:1.1.2'
     androidTestImplementation 'androidx.test.espresso:espresso-core:3.3.0'
+
```
Отформатируйте свой ПР пожалуйста. Удалите лишние строки. Примените автоформат в студии.

elestatte87 commented on app/src/main/AndroidManifest.xml:
```diff
@@ -8,7 +9,13 @@
-        android:theme="@style/Theme.RobertAndroidCourse2021">
+        android:theme="@style/AppTheme"
+        tools:ignore="AllowBackup,GoogleAppIndexingWarning">
+        <service
```
удалите ненужные таги, это те что по-дефолту установлены в нужное значение или наследуются от рута (application)

elestatte87 commented on app/src/main/java/com/example/robertandroidcourse2021/ContactDetailsFragment.kt:
```diff
@@ -0,0 +1,46 @@
+    companion object {
+        fun getNewInstance(id: Int) =
+            ContactDetailsFragment().apply {
+                arguments = Bundle().apply {
```
вы уже подключили ktx - используйте bundleOf

elestatte87 commented on app/src/main/java/com/example/robertandroidcourse2021/ContactDetailsFragment.kt:
```diff
@@ -0,0 +1,46 @@
+        val service = (activity as?  ServiceProvider)?.getService()!!
+        val contact = service.getContactDetails(id)
+
+        binding.contactImageDetail.setImageResource(contact.imageResId)
```
with (binding) {}

elestatte87 commented on app/src/main/java/com/example/robertandroidcourse2021/ContactService.kt:
```diff
@@ -0,0 +1,39 @@
+        return getFakeContacts()
+    }
+
+    fun getContactDetails(id : Int):ContactModel{
```
Мы обсуждали этот момент на лекции и в чате. Необходимо сымитировать долгую загрузку данных введя задержку небольшую на получение (1000-1500 мс). тут у вас свобода действий нативно или коорутины

elestatte87 commented on app/src/main/java/com/example/robertandroidcourse2021/ContactService.kt:
```diff
@@ -0,0 +1,39 @@
+import android.os.IBinder
+
+
+/*interface ContactProvider{
```
закоменченного кода быть не должно

elestatte87 commented on app/src/main/java/com/example/robertandroidcourse2021/ContactService.kt:
```diff
@@ -0,0 +1,39 @@
+    }
+
+    fun getContactDetails(id : Int):ContactModel{
+        return getFakeContacts().find{it.id == id}!!
```
!! - будет краш если не найдет контакт. Стоит обработать этот кейс по-другому. Через опциональность и пока чего-то пользователю
Старайтесь обходиться опциональностью и обработкой null
а не !! и lateinit

elestatte87 commented on app/src/main/java/com/example/robertandroidcourse2021/ContactService.kt:
```diff
@@ -0,0 +1,39 @@
+        return binder
+
+    }
+//чтобы получить объект ContactService
```
Можно без коментариев

elestatte87 commented on app/src/main/res/drawable/ic_launcher_background.xml:
```diff
@@ -1,170 +1,199 @@
-<?xml version="1.0" encoding="utf-8"?>
+<?xml version="1.0" encoding="utf-8"?><!--
```
не делайте пожалуйста такой диф - это усложняет ревью кода
это уже второй раз откатите это изменение

elestatte87 commented on app/src/main/java/com/example/robertandroidcourse2021/ContactListFragment.kt:
```diff
@@ -0,0 +1,46 @@
+
+    companion object{
+        fun getNewInstance() =
+            ContactListFragment().apply {
```
Нарушается принцип YAGNI

elestatte87 commented on app/src/main/java/com/example/robertandroidcourse2021/ContactListFragment.kt:
```diff
@@ -0,0 +1,46 @@
+import android.view.ViewGroup
+import androidx.fragment.app.Fragment
+import com.example.robertandroidcourse2021.databinding.FragmentContactListBinding
+import kotlinx.android.synthetic.main.fragment_contact_list.*
```
у вас подключен вью биндинг - котлин синтетик не нужен теперь

elestatte87 commented on app/src/main/java/com/example/robertandroidcourse2021/ContactListFragment.kt:
```diff
@@ -0,0 +1,46 @@
+        val contacts = service.getContactList()
+        val contact = contacts[0]
+        contact_image.setImageResource(contact.imageResId)
+        binding.name.text = contact.name
```
with(binding) {
}

elestatte87 commented on app/src/main/java/com/example/robertandroidcourse2021/ContactListFragment.kt:
```diff
@@ -0,0 +1,46 @@
+        super.onViewCreated(view, savedInstanceState)
+
+        val service = (activity as?  ServiceProvider)?.getService()!!
+        val contacts = service.getContactList()
```
синхронный вызов зафризит ui

robertis93 committed:
bindservice without comments

elestatte87 commented on app/src/main/res/values/strings.xml:
```diff
@@ -1,3 +1,7 @@
\ No newline at end of file
+  <string name="app_name">RobertAndroidCourse</string>
+  <string name="description_format">%1$s\n\n%2$s</string>
+  <string name="contactdetailsfragment">ContactDetailsFragment</string>
```
не соотвествует соглашениям по неймингу. Snake case в ресурсах. Это абсолютно не читаемо

elestatte87 commented on app/src/main/res/values/styles.xml:
```diff
@@ -0,0 +1,11 @@
+<?xml version="1.0" encoding="utf-8"?>
```
зачем вы удалили theme и перенеесли все сюда. В этом нет смысла
в стили обычно выносят стили компонентов шрифтов.
Не стоит делать ненужных вещей, 50% изменений не имеют отношения к поставленной задаче

elestatte87 commented on app/src/main/res/layout/fragment_contact_list.xml:
```diff
@@ -0,0 +1,66 @@
+            android:textSize="24sp" />
+    </androidx.appcompat.widget.Toolbar>
+
+    <androidx.constraintlayout.widget.ConstraintLayout
```
у вас ConstraintLayout в ConstraintLayout это worst practice
Если хотите ConstraintLayout то тогда 1 - в этом его суть
либо вообще без него обойдитесь верстка тут простая.

elestatte87 replied:
а зачем вам теперь вообще тут лишний лейаут удалите и его

elestatte87 replied:
разобрался как вы себя загнали в такую ситуацию

elestatte87 commented on app/src/main/res/layout/activity_main.xml:
```diff
@@ -1,7 +1,11 @@
     android:layout_height="match_parent"
-    tools:context=".MainActivity"/>
\ No newline at end of file
+    android:paddingLeft="@dimen/recycler_margin_horizontal"
```
paddingHorizontal
paddingVertical

elestatte87 commented on app/src/main/res/layout/activity_main.xml:
```diff
@@ -1,7 +1,11 @@
+    android:paddingRight="@dimen/recycler_margin_horizontal"
+    android:paddingBottom="@dimen/recycler_margin_vertical">
+
+</FrameLayout>
```
Закрывающийся таг можно не писать - если сингл таг
<FrameLayout ..... />

elestatte87 commented on app/src/main/res/layout/fragment_contact_details.xml:
```diff
@@ -0,0 +1,101 @@
+            android:layout_width="match_parent"
+            android:layout_height="wrap_content"
+            android:orientation="vertical">
+            <androidx.appcompat.widget.Toolbar
```
Не видел ни одного приложения где тулбар скроллится. Вынесите его из скролла.
Все коментарии анализируйте по-всему ПР

elestatte87 commented on app/src/main/res/layout/fragment_contact_list.xml:
```diff
@@ -0,0 +1,66 @@
+        app:layout_constraintStart_toStartOf="parent"
+        app:layout_constraintTop_toBottomOf="@id/toolbar">
+
+        <ImageView
```
Вот это все лучше сделать в отдельном файле а тут сделать инклуд в рут контейнер фрагмента, так как потом это будет ячейка списка
Что стало со старой версткой ? Тут тяжело соориентироваться вижу новую верстку, не видно что стало со старой. Вы же уже все это делали

elestatte87 commented on app/src/main/res/layout/fragment_contact_list.xml:
```diff
@@ -0,0 +1,66 @@
+            android:id="@+id/contactNumber"
+            android:layout_width="wrap_content"
+            android:layout_height="wrap_content"
+            android:text="@{contactModel.numberOne}"
```
через код все данные устаналивайте потом поймете почему в след ДЗ

elestatte87 commented on app/src/main/java/com/example/robertandroidcourse2021/MainActivity.kt:
```diff
@@ -1,11 +1,80 @@
+
+    lateinit var contactService: ContactService
+
+    private var isBounded = false
```
где анализируется это поле ? (оно должно анализироваться)
Например чтобы не биндиться если уже есть бинд. Также во фрагментах вам надо будет дожидаться пока не закончиться бинд сервиса и потом грузить данные.
Сейчас не понятно как вы с ним оперируете.

elestatte87 commented on app/src/main/java/com/example/robertandroidcourse2021/MainActivity.kt:
```diff
@@ -1,11 +1,80 @@
+
+        override fun onServiceConnected(name: ComponentName?, service: IBinder?) {
+            val binder = service as ContactService.ContactServiceBinder
+            GlobalScope.launch { contactService = binder.getService()
```
почему GlobalScope ? почитайте про коорутины
https://kotlin.github.io/kotlinx.coroutines/kotlinx-coroutines-core/kotlinx.coroutines/-global-scope/
зачем вам именно тут коорутины. Примените их по делу например для имитации загрузки данных

robertis93 committed:
bindservice fixed

elestatte87 commented on app/src/main/java/com/example/robertandroidcourse2021/ContactDetailsFragment.kt:
```diff
@@ -0,0 +1,74 @@
+
+class ContactDetailsFragment : Fragment() {
+
+    lateinit var binding: FragmentContactDetailsBinding
```
private

elestatte87 commented on app/src/main/java/com/example/robertandroidcourse2021/ContactListFragment.kt:
```diff
@@ -0,0 +1,58 @@
+            val service = (activity as? ServiceProvider)?.getService()
+            val contacts = service?.getContactList()
+            withContext(Dispatchers.Main) {
+                val contact = contacts?.get(0)
```
val contact = contacts?.get(0) ?: return
и уберите все if

elestatte87 commented on app/src/main/java/com/example/robertandroidcourse2021/ContactListFragment.kt:
```diff
@@ -0,0 +1,58 @@
+
+    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
+        super.onViewCreated(view, savedInstanceState)
+        CoroutineScope(Dispatchers.IO).launch {
```
необходимо отменять все коорутины в detachView

elestatte87 commented on app/src/main/java/com/example/robertandroidcourse2021/ContactService.kt:
```diff
@@ -0,0 +1,32 @@
+    }
+
+    suspend fun getContactList(): List<ContactModel> {
+
```
тоже delay

elestatte87 commented on app/src/main/java/com/example/robertandroidcourse2021/MainActivity.kt:
```diff
@@ -1,11 +1,64 @@
         super.onCreate(savedInstanceState)
         setContentView(R.layout.activity_main)
+
+        val fragment = ContactListFragment.getNewInstance()
```
под if занесите, понятно почему ?

elestatte87 commented on app/src/main/java/com/example/robertandroidcourse2021/ContactDetailsFragment.kt:
```diff
@@ -0,0 +1,74 @@
+    companion object {
+        fun getNewInstance(id: Int) =
+            ContactDetailsFragment().apply {
+                arguments = bundleOf(Pair("contactId", id))
```
есть инфиксный оператор

elestatte87 commented on app/src/main/java/com/example/robertandroidcourse2021/MainActivity.kt:
```diff
@@ -1,11 +1,64 @@
+    override fun onStart() {
+        super.onStart()
+        val intent = Intent(this, ContactService::class.java)
+        bindService(intent, connection, Context.BIND_AUTO_CREATE)
```
слишком часто это будет дергаться лучше по реже в более ранних колбеках
isBinded зря удалили оно должно анализироваться во фрагментах так как биндинг происходит дольше чем отображение фрагмента

robertis93 committed:
bindservice fixed

robertis93 committed:
bindservice fixed

robertis93 committed:
bindservice fixed

elestatte87 commented on app/src/main/java/com/example/robertandroidcourse2021/ContactDetailsFragment.kt:
```diff
@@ -0,0 +1,51 @@
+    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
+        super.onViewCreated(view, savedInstanceState)
+        val id = arguments!!.getInt("contactId")
+        if (MainActivity.mBound)
```
Вот это прямо плохое решение статик поле в активити
у вас же есть интерфейс сделайте там еще 1 метод
val service = (activity as? ServiceProvider)?.getService()
service?.isBound()

elestatte87 commented on app/src/main/java/com/example/robertandroidcourse2021/ContactDetailsFragment.kt:
```diff
@@ -0,0 +1,51 @@
+        super.onViewCreated(view, savedInstanceState)
+        val id = arguments!!.getInt("contactId")
+        if (MainActivity.mBound)
+            lifecycleScope.launch {
```
Кажется у вас тут работа теперь будет идти на мейн треде.
Сделайте ио скоуп  в нем запрашивайте данные, этот скоуп создавайте - отмеяйте и обнуляйте в аттаче - детаче
 UI обновляйте на UI треде. Почитайте доку по коорутинам, это стандартная тема там 100% есть решения как работать с ними в такой ситуации

elestatte87 commented on app/src/main/java/com/example/robertandroidcourse2021/ContactListFragment.kt:
```diff
@@ -0,0 +1,45 @@
+
+    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
+        super.onViewCreated(view, savedInstanceState)
+        if (MainActivity.mBound)
```
Все тоже что в другом фрашменте

elestatte87 commented on app/src/main/java/com/example/robertandroidcourse2021/ContactModel.kt:
```diff
@@ -0,0 +1,14 @@
+    val description: String
+)
+
+
```
ладно если с форматированием никак то ок. Но это минус в зачет

elestatte87 commented on app/src/main/java/com/example/robertandroidcourse2021/MainActivity.kt:
```diff
@@ -1,11 +1,67 @@
+class MainActivity : AppCompatActivity(), ServiceProvider {
+
+    companion object {
+        var mBound: Boolean = false
```
Удалите это - это плохая практика статик поля в активити

elestatte87 commented on app/src/main/java/com/example/robertandroidcourse2021/MainActivity.kt:
```diff
@@ -1,11 +1,67 @@
+                .commit()
+
+            val intent = Intent(this, ContactService::class.java)
+            bindService(intent, connection, Context.BIND_AUTO_CREATE)
```
нет смысла делать биндинг если mBound == true
и анбиндинг если mBound == false

elestatte87 commented on app/src/main/java/com/example/robertandroidcourse2021/MainActivity.kt:
```diff
@@ -1,11 +1,67 @@
+        var mBound: Boolean = false
+    }
+
+    override fun getService() = ContactService()
```
Вы же понимаете что это неправильно ?
сделайте поле contactService в котором будете хранить сервис и присвойте этой переменной значение в connection в onServiceConnected

val binder = (service as? ContactService.ContactBinder) ?: return
contactService = binder.getService()

override fun getService() = contactService

elestatte87 commented on app/src/main/res/layout/contact_inn.xml:
```diff
@@ -0,0 +1,38 @@
+<?xml version="1.0" encoding="utf-8"?>
+<androidx.constraintlayout.widget.ConstraintLayout
```
Ок вот вам план сделайте тут не констрейнт (тут без него можно сверстать)
После этого сможете избавиться от экстра констрейнт лейаута в верстке фрагмента и сделать сразу инклуд этого лейаута - это решит все ваши проблемы с версткой в этом ПР

elestatte87 commented on app/src/main/res/values/dimens.xml:
```diff
@@ -0,0 +1,12 @@
+<?xml version="1.0" encoding="utf-8"?>
+<resources>
+
```
Вообщем можете игнорировать мои просьбы о форматировании - просто минус к оценке

robertis93 committed:
notification/ alarm manager

robertis93 committed:
with error "FATAL EXCEPTION: DefaultDispatcher-worker-1
                Process: com.example.robertandroidcourse2021, PID: 15315
                kotlin.UninitializedPropertyAccessException: lateinit property contactService has not been initialized"

robertis93 committed:
Merge remote-tracking branch 'origin/ibind_hw' into ibind_hw

# Conflicts:
#	app/src/main/java/com/example/robertandroidcourse2021/ContactDetailsFragment.kt
#	app/src/main/java/com/example/robertandroidcourse2021/ContactListFragment.kt

robertis93 committed:
with error "FATAL EXCEPTION: DefaultDispatcher-worker-1
                Process: com.example.robertandroidcourse2021, PID: 15315
                kotlin.UninitializedPropertyAccessException: lateinit property contactService has not been initialized"

robertis93 committed:
23.04

robertis93 committed:
without mistakes

robertis93 committed:
without mistakes

elestatte87 commented on app/src/main/java/com/example/robertandroidcourse2021/ContactListFragment.kt:
```diff
@@ -0,0 +1,57 @@
+        if (isBound) {
+            lifecycleScope.launch(IO) {
+                val service = (activity as? ServiceProvider)?.getService()
+                val contacts = service?.getContactList()
```
надо дождаться тут пока сервис будет забинжен

elestatte87 commented on app/src/main/java/com/example/robertandroidcourse2021/ContactDetailsFragment.kt:
```diff
@@ -0,0 +1,57 @@
+    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
+        super.onViewCreated(view, savedInstanceState)
+        val id = arguments!!.getInt("contactId")
+        val isBound = (activity as ServiceProvider).getBound()
```
надо дождаться тут когда сервис будет забинжен он может тут быть и не забинжен еще

elestatte87 replied:
не сделано можете просто через while дождаться

elestatte87 commented on app/src/main/java/com/example/robertandroidcourse2021/ContactListFragment.kt:
```diff
@@ -0,0 +1,57 @@
+
+    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
+        super.onViewCreated(view, savedInstanceState)
+        val isBound = (activity as ServiceProvider).getBound()
```
надо дождаться тут когда сервис будет забинжен он может тут быть и не забинжен еще.
еще старайтесь минимизировать вложенность кода там где это очевидно.
if (...) return
blabla
вместо
if (...) {
    blabla
}

elestatte87 replied:
надо дождаться тут когда сервис будет забинжен он может тут быть и не забинжен еще. - не сделано

elestatte87 commented on app/src/main/java/com/example/robertandroidcourse2021/MainActivity.kt:
```diff
@@ -1,11 +1,66 @@
+
+class MainActivity : AppCompatActivity(), ServiceProvider {
+    private var mBound = false
+    private lateinit var contactService: ContactService
```
лучше private var contactService: ContactService? = null

elestatte87 commented on app/src/main/java/com/example/robertandroidcourse2021/MainActivity.kt:
```diff
@@ -1,11 +1,66 @@
+
+    override fun onDestroy() {
+        super.onDestroy()
+        unbindService(connection)
```
if (mBound ) {
unbind
mBound = false
}
делать это надо до super.onDestroy()

elestatte87 commented on app/src/main/java/com/example/robertandroidcourse2021/MainActivity.kt:
```diff
@@ -1,11 +1,66 @@
         super.onCreate(savedInstanceState)
         setContentView(R.layout.activity_main)
+        val intent = Intent(this, ContactService::class.java)
+        bindService(intent, connection, Context.BIND_AUTO_CREATE)
```
if (!mBound ) {
val intent = Intent(this, ContactService::class.java)
bindService(intent, connection, Context.BIND_AUTO_CREATE)
}

robertis93 committed:
changed the last comments

elestatte87 commented on app/src/main/java/com/example/robertandroidcourse2021/ContactListFragment.kt:
```diff
@@ -0,0 +1,54 @@
+        super.onViewCreated(view, savedInstanceState)
+        lifecycleScope.launch(IO) {
+            val isBound = (activity as ServiceProvider).getBound()
+            if (!isBound) {
```
Что это вам дает ? Какова практическая ценность этого кода ?

elestatte87 commented on app/src/main/java/com/example/robertandroidcourse2021/ContactListFragment.kt:
```diff
@@ -0,0 +1,54 @@
+    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
+        super.onViewCreated(view, savedInstanceState)
+        lifecycleScope.launch(IO) {
+            val isBound = (activity as ServiceProvider).getBound()
```
тут (activity as ServiceProvider) ниже (activity as? ServiceProvider)

elestatte87 commented on app/src/main/java/com/example/robertandroidcourse2021/ContactDetailsFragment.kt:
```diff
@@ -0,0 +1,57 @@
+        super.onViewCreated(view, savedInstanceState)
+        val id = arguments!!.getInt("contactId")
+        lifecycleScope.launch(IO) {
+            val isBound = (activity as ServiceProvider).getBound()
```
тут (activity as ServiceProvider) ниже (activity as? ServiceProvider)

elestatte87 commented on app/src/main/java/com/example/robertandroidcourse2021/ContactDetailsFragment.kt:
```diff
@@ -0,0 +1,57 @@
+        val id = arguments!!.getInt("contactId")
+        lifecycleScope.launch(IO) {
+            val isBound = (activity as ServiceProvider).getBound()
+            if (isBound) {
```
дождитесь в цикле пока сервис будет забинжен
